File: scripts/build-pinyin-audio.mjs
#!/usr/bin/env node
/**
 * Download pinyin teaching audio:
 * - isolated initials/finals: zxq432/py d/py/ (b.mp3, a1.mp3, …)
 * - tone examples: mp3-chinese-pinyin-sound (ma1.mp3, …)
 *
 * Usage: node scripts/build-pinyin-audio.mjs [--force]
 */

import { readdirSync, unlinkSync, mkdirSync, readFileSync, writeFileSync, existsSync, statSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');
const AUDIO_DIR = join(ROOT, 'audio', 'pinyin');
const PINYIN_JS = join(ROOT, 'js', 'pinyin.js');
const MANIFEST_OUT = join(AUDIO_DIR, 'manifest.json');

const ISOLATED_BASE =
  'https://raw.githubusercontent.com/zxq432/py/main/d/py/{key}.mp3';
const SYLLABLE_BASE =
  'https://raw.githubusercontent.com/davinfifield/mp3-chinese-pinyin-sound/master/mp3/{key}.mp3';
const SYLLABLE_FALLBACK =
  'https://raw.githubusercontent.com/hugolpz/audio-cmn/master/64k/syllabs/cmn-{key}.mp3';

const force = process.argv.includes('--force');

function parseManifest() {
  const text = readFileSync(PINYIN_JS, 'utf8');
  const entries = new Map();
  const re =
    /\{[^{}]*audio:\s*'([^']+\.mp3)'[^{}]*audioKey:\s*'([^']+)'[^{}]*audioSource:\s*'([^']+)'/g;
  let m;
  while ((m = re.exec(text)) !== null) {
    const [, filename, audioKey, audioSource] = m;
    if (!entries.has(filename)) {
      entries.set(filename, { filename, audioKey, audioSource });
    }
  }
  return [...entries.values()];
}

async function tryDownload(url) {
  const res = await fetch(url);
  if (!res.ok) return null;
  const buf = Buffer.from(await res.arrayBuffer());
  return buf.length > 300 ? buf : null;
}

async function downloadItem(item, dest) {
  if (item.audioSource === 'isolated') {
    const data = await tryDownload(ISOLATED_BASE.replace('{key}', item.audioKey));
    if (data) {
      writeFileSync(dest, data);
      return 'zxq432-py-isolated';
    }
    return null;
  }

  for (const [url, source] of [
    [SYLLABLE_BASE.replace('{key}', item.audioKey), 'mp3-chinese-pinyin-sound'],
    [SYLLABLE_FALLBACK.replace('{key}', item.audioKey), 'audio-cmn-syllabs']
  ]) {
    const data = await tryDownload(url);
    if (data) {
      writeFileSync(dest, data);
      return source;
    }
  }
  return null;
}

function cleanOldAudio(keep) {
  if (!existsSync(AUDIO_DIR)) return;
  for (const file of readdirSync(AUDIO_DIR)) {
    if (!file.endsWith('.mp3')) continue;
    if (!keep.has(file)) unlinkSync(join(AUDIO_DIR, file));
  }
}

async function main() {
  mkdirSync(AUDIO_DIR, { recursive: true });
  const items = parseManifest();
  const keep = new Set(items.map(i => i.filename));
  if (force) cleanOldAudio(keep);

  console.log(`Building ${items.length} pinyin audio files -> ${AUDIO_DIR}`);

  const manifest = [];
  const failed = [];

  for (const item of items) {
    const dest = join(AUDIO_DIR, item.filename);
    let source = 'existing';

    if (!force && existsSync(dest) && statSync(dest).size > 300) {
      source = 'existing';
    } else {
      source = (await downloadItem(item, dest)) || 'missing';
      if (source === 'missing') failed.push(`${item.filename} (${item.audioKey})`);
    }

    manifest.push({ ...item, source });
  }

  const ok = manifest.filter(
    e => existsSync(join(AUDIO_DIR, e.filename)) && statSync(join(AUDIO_DIR, e.filename)).size > 300
  ).length;

  writeFileSync(
    MANIFEST_OUT,
    JSON.stringify(
      {
        total: manifest.length,
        ok,
        failed,
        note: 'Initials/finals: isolated clips; tones: ma syllable clips',
        files: manifest,
        attribution: {
          isolated: 'zxq432/py d/py/ — decomposed pinyin phonetics',
          syllable: 'davinfifield/mp3-chinese-pinyin-sound (Unlicense)'
        }
      },
      null,
      2
    ),
    'utf8'
  );

  console.log(`Done: ${ok}/${manifest.length} files ready`);
  if (failed.length) {
    console.error('Missing:', failed.join(', '));
    process.exit(1);
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: scripts/build-pinyin-audio.py
#!/usr/bin/env python3
"""Download dedicated pinyin syllable MP3s (NOT HSK word readings)."""

from __future__ import annotations

import json
import re
import sys
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
AUDIO_DIR = ROOT / "audio" / "pinyin"
PINYIN_JS = ROOT / "js" / "pinyin.js"
MANIFEST_OUT = AUDIO_DIR / "manifest.json"
PINYIN_MP3_BASE = (
    "https://raw.githubusercontent.com/davinfifield/mp3-chinese-pinyin-sound/master/mp3/{key}.mp3"
)
SYLLABS_BASE = (
    "https://raw.githubusercontent.com/hugolpz/audio-cmn/master/64k/syllabs/cmn-{key}.mp3"
)


def parse_manifest_from_js() -> list[dict]:
    text = PINYIN_JS.read_text(encoding="utf-8")
    entries: dict[str, dict] = {}
    for match in re.finditer(r"\{[^{}]*audio:\s*'([^']+\.mp3)'[^{}]*audioKey:\s*'([^']+)'", text):
        filename, audio_key = match.groups()
        if filename not in entries:
            entries[filename] = {"filename": filename, "audioKey": audio_key}
    return list(entries.values())


def try_download(url: str) -> bytes | None:
    try:
        with urllib.request.urlopen(url, timeout=30) as resp:
            data = resp.read()
        return data if len(data) >= 500 else None
    except (urllib.error.HTTPError, urllib.error.URLError, TimeoutError):
        return None


def download_pinyin_syllable(audio_key: str, dest: Path) -> str | None:
    for url, source in (
        (PINYIN_MP3_BASE.format(key=audio_key), "mp3-chinese-pinyin-sound"),
        (SYLLABS_BASE.format(key=audio_key), "audio-cmn-syllabs"),
    ):
        data = try_download(url)
        if data:
            dest.write_bytes(data)
            return source
    return None


def main() -> int:
    force = "--force" in sys.argv
    AUDIO_DIR.mkdir(parents=True, exist_ok=True)
    items = parse_manifest_from_js()
    keep = {item["filename"] for item in items}

    if force:
        for path in AUDIO_DIR.glob("*.mp3"):
            if path.name not in keep:
                path.unlink()

    print(f"Building {len(items)} pinyin syllable files -> {AUDIO_DIR}")

    manifest: list[dict] = []
    failed: list[str] = []

    for item in items:
        dest = AUDIO_DIR / item["filename"]
        source = "existing"

        if not force and dest.exists() and dest.stat().st_size > 500:
            source = "existing"
        else:
            downloaded = download_pinyin_syllable(item["audioKey"], dest)
            if downloaded:
                source = downloaded
            else:
                source = "missing"
                failed.append(f"{item['filename']} ({item['audioKey']})")

        manifest.append({**item, "source": source})

    ok = sum(
        1
        for m in manifest
        if (AUDIO_DIR / m["filename"]).exists()
        and (AUDIO_DIR / m["filename"]).stat().st_size > 500
    )

    MANIFEST_OUT.write_text(
        json.dumps(
            {
                "total": len(manifest),
                "ok": ok,
                "failed": failed,
                "note": "Dedicated pinyin syllable recordings — not HSK character/word audio",
                "files": manifest,
                "attribution": {
                    "primary": "davinfifield/mp3-chinese-pinyin-sound (Unlicense)",
                    "fallback": "hugolpz/audio-cmn syllabs (CC-BY-SA, Chen Wang)",
                },
            },
            ensure_ascii=False,
            indent=2,
        ),
        encoding="utf-8",
    )

    print(f"Done: {ok}/{len(manifest)} files ready")
    if failed:
        print("Missing:", ", ".join(failed), file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
